Add Bookmark Folders

THISISDINOSAUR opened the pull request:
<!--
Note: This checklist is a reminder of our shared engineering expectations. Feel free to change it, although assigning a GitHub reviewer and the items in bold are required.
-->

Task/Issue URL: https://app.asana.com/0/72649045549333/1143141356446970/f
Tech Design URL: https://app.asana.com/0/481882893211075/1200283968383818/f, https://app.asana.com/0/481882893211075/1201023911968241/f
CC:

**Description**:
Adds bookmark folders, including migrating bookmark and favourite storage from UserDefaults to CoreData.

**Steps to test this PR**:
Bookmarks/favorites appear in many parts of the app, so the test plan itself below should be scrutinised for anything missing. I've broken it down by area of the app to hopefully make that easier

**Test Migration**
1. To start, before you run this build, make sure you have favorites and bookmarks on a previous version of the app, and screenshot them, so you can compare their existence/position both pre and post migration
2. Install this PR
3. Open the app and go to the bookmark screen and check all favourites and bookmarks are in the right place (their ordering should be preserved)
 ---
1. Start with a previous build of the app again, with favs and bookmarks 
2. Install this PR 
3. Before you go to the bookmarks screen, add some favorites and bookmarks
4. Then go to the bookmarks screen, and check previous favs and bookmarks are in the right positions, and any new ones were correctly added to the end.
 ---
1. Delete any existing DDG app
2. Install this PR
3. Go to the bookmarks screen, ensure there is nothing there
4. Add a folder/bookmark/whatever, check it works correctly

**The browsing menu**
1. Go to a website that isn't a fav or bookmark
2. Check that the browsing menu shows the options to add it as a bookmark and a favorite
3. Add it as a bookmark from the browsing menu
4. Add it as a favourite from the browsing menu
5. Now when opening the browsing menu it should show the options to edit the bookmark, and remove the favorite
6. Edit the bookmark from the browsing menu, make sure the edit screen is the correct one for the bookmark and not the favourite.
7. Make some change (e.g. to the title) and save
8. Go to the bookmark screen to check the change saved correctly 
9. Go back to the browsing menu and remove the favourite
10. Open the browsing menu and check that the remove favourite option has changed to add favourite, and the bookmark option still says edit bookmark.
 ---
1. Create a bookmark from the browsing menu again
2. When the toast appears at the bottom of the screen to say the bookmark is created, press the edit button
3. Edit the bookmark and check it works as above
---
1. Create a favorite from the browsing menu 
2. When the toast appears, press undo. Check that the favorite doesn't exist
3. Do the reverse (remove a favourite from the browsing menu and press undo)

**The bookmarks screen**
1. Check that bookmarks and favourites show the correct favicons
2. Press on both a favorite and a bookmark, check they open the sites correctly
3. Swipe right on both a bookmark and a favorite and share them
4. Swipe left on a bookmark, a favorite and a folder and delete them. Check an empty folder doesn't show the confirmation prompt
5. Attempt to delete a folder with contents, it should show a confirmation prompt, check both options work as expected.
6. Press on a folder and check it shows the correct subfolder screen. 
---
1. Press the edit button, the cells should change to show the edit view
2. Rearrange some bookmarks and favorites 
3. Rearrange a bookmark into a favourite and vice versa
4. Attempting to do likewise with a folder, it shouldn't let you convert it to a folder.
5. Also test arranging bookmarks/folders from a sub folder.
---
1. When there are no bookmarks or favorites, check it shows the correct empty state
2. When a subfolder has no items, check it shows the correct empty state (different from the top level empty state)
---
1. Edit a bookmark, select a new parent folder
2. Save it, and check all views etc updated correctly 
---
1. Edit a bookmark, but press add a new folder
2. After creating a new folder, it should be selected automatically yon the edit bookmark screen
---
1. Edit a favorite, checks it doesn't show the select folder view
2. Make some changes and test saving it works 
---
1. Edit a favorite or bookmark, but don't save the changes. Check that no changes are saved
---
1. Edit a folder
2. Check that in the select folder view itself or any of its children don't appear (i.e. you shouldn't be able to add a folder to itself)

**Favorites in various places**
1. On iPad, check that the favourites overlay appears and its contents is still correct when using the url bar
---
1. Check that favorutes appear on the home screen
2. Long press and check that both editing and deleting favorites works from the home screen.
3. Whist on the Home Screen, open the bookmarks view and delete or edit a favorite, check this change is reflected in the Home Screen view

**Search**
1. From the bookmarks screen, scroll down to reveal the search bar
2. Try search terms that should show a) a bookmark b) a favorite, c) both, d) neither, and check the results are correct
3. Tap on a result to check that you can open a bookmark from the search results
4. Try to swipe to delete a search result, it shouldn't work
5. Try to swipe to share a search result, it _should_ work
Note: Folders should not appear in search results 
---
1. Type in URL bar search terms as above, check they are correct and the sites open correctly.


**Widgets**
1. When you have favorites already, add either the medium or large widget (i.e. one with favorites)
2. Check that the favorites appear there as expected (in the right order, with the right favicons etc)
3. Tap on a favourite and check it opens the site as expected
---
1. When you have no favorites and have not already been through favorites onboarding, add the widget
2. It should show the onboarding prompt, press on it
3. The app should show the favourites onboarding 
4. Once you've added a favorite, check the widget shows the new favorite
5. Then delete the fav, and check the change is propagated to the widget
---
1. Convert a bookmark to a favorites, or vice versa, and check the change is propagated to the widget

**Other**
1. Press the bookmark all tabs button from tabs screen, check that it does that.

**Pixels**
Test the following pixels fire appropriately
- case bookmarksFolderCreated = "m_bookmarks_folder_created"
- case bookmarkCreatedAtTopLevel = "m_bookmark_created_at_top_level"
- case bookmarkCreatedInSubfolder = "m_bookmark_created_in_subfolder"
- case bookmarkEditedAtTopLevel = "m_bookmark_edited_at_top_level"
- case bookmarkEditedInSubfolder = "m_bookmark_edited_in_subfolder"

<!--
Before submitting a PR, please ensure you have tested the combinations you expect the reviewer to test, then delete configurations you *know* do not need explicit testing.

Using a simulator where a physical device is unavailable is acceptable.
-->

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [x] Portrait
* [x] Landscape

**Device Testing**:

* [x] iPhone SE (1st Gen)
* [ ] iPhone 8
* [x] iPhone X
* [x] iPad

**OS Testing**:

* [ ] iOS 13
* [x] iOS 14
* [x] iOS 15

**Theme Testing**:

* [x] Light theme
* [x] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
**When ready for review, remember to post the PR in MM**

THISISDINOSAUR committed:
Initial commit

THISISDINOSAUR committed:
Clean up unused bookmark code

THISISDINOSAUR committed:
Make bookmark URL nonoptional

THISISDINOSAUR committed:
Add cunning code I'm unlikely to use

THISISDINOSAUR committed:
Add much smarter method of handling folder expansion and indexing cells

THISISDINOSAUR committed:
Add cell styling for folders/new bookmark object

THISISDINOSAUR committed:
Change data source to use new core data store for bookmarks

THISISDINOSAUR committed:
Update storyboard

THISISDINOSAUR committed:
Make favourites powered by new data store

THISISDINOSAUR committed:
Add bar items to BookmarksViewController

THISISDINOSAUR committed:
Add toolbar theming

THISISDINOSAUR committed:
Fix canEdit check for bookmark cells

THISISDINOSAUR committed:
Add check to stop folders being moved to favourites

THISISDINOSAUR committed:
Remove unused data source methods

THISISDINOSAUR committed:
Remove unused file

THISISDINOSAUR committed:
Add new favourites to new bookmark screen and make favourites display optional
Also lay groundwork for having variable parent folder

THISISDINOSAUR committed:
Update bookmark selection logic

THISISDINOSAUR committed:
Add child folder view

THISISDINOSAUR committed:
Fix bookmark selection from child folders

THISISDINOSAUR committed:
Remove unnecessary pre iOS 13 code

THISISDINOSAUR committed:
Make search bar not appear on child folders

THISISDINOSAUR committed:
Fix gap in viewcontroller when transitioning to child folder

THISISDINOSAUR committed:
Add initial Add folder screen

THISISDINOSAUR committed:
Remove unused dependancy from BookmarksViewController

THISISDINOSAUR committed:
Add separate tableViewController for folders

THISISDINOSAUR committed:
Add BookmarkFolderCell styling

THISISDINOSAUR committed:
Add new data source for folder trees

THISISDINOSAUR committed:
Refactor bookmark tableview data sources so each section has a separate data source
And so any table view can have an arbitrary mix of sections

THISISDINOSAUR committed:
Add bookmark folders section data source

THISISDINOSAUR committed:
Add per section commit editing style

THISISDINOSAUR committed:
Add Bookmark details data source and view to new folder screen

THISISDINOSAUR committed:
Fix top level bookmarks folder title on Add folder screen

THISISDINOSAUR committed:
Add translation layer between core data and rest of app

THISISDINOSAUR committed:
Add new folder creation to data store

THISISDINOSAUR committed:
Fix folder saving not being performed in context.perform

THISISDINOSAUR committed:
Add folder updating

THISISDINOSAUR committed:
Rename AddBookmarksFolderViewController to reflect it also supports editing
Also make navigation title change based on edit v add

THISISDINOSAUR committed:
Add bookmarks edit screen

THISISDINOSAUR committed:
Add bookmark details cell

miasma13 committed:
Clean up naming of the dev & ad hoc provisioning profiles (#973)

* Update names of the ad hoc build provisioning profiles

* Update the name of development provisioning profile

THISISDINOSAUR committed:
Fix bookmark editing from main screen

THISISDINOSAUR committed:
Replace bookmark and favorite detail views

THISISDINOSAUR committed:
Fix section titles

THISISDINOSAUR committed:
Fix various styling

THISISDINOSAUR committed:
Disable sharing on folders

THISISDINOSAUR committed:
Fix entire bookmarks view dismissing after completing edit screen

THISISDINOSAUR committed:
Split edit fav, folder, bookmark into seperate data sources

THISISDINOSAUR committed:
Add separate titles for edit/add bookmarks and folders

THISISDINOSAUR committed:
Fix edit bookmark cell styling

THISISDINOSAUR committed:
Update save/update functions

THISISDINOSAUR committed:
Fix items rearranging on update if the parent doesn't change

THISISDINOSAUR committed:
Add core data rearranging functions

THISISDINOSAUR committed:
Update delete function

THISISDINOSAUR committed:
Add convert to fav/bookmark methods

THISISDINOSAUR committed:
Huge core data layer refactor to make everything async, safe and use
two separate contexts for editing v viewing

THISISDINOSAUR committed:
Fix update on data change

THISISDINOSAUR committed:
Fix favourites empty state showing bookmarks empty state

THISISDINOSAUR committed:
Fix deletion sometimes crashing

THISISDINOSAUR committed:
Fix bookmark rearranging

THISISDINOSAUR committed:
Make search use coredata

THISISDINOSAUR committed:
Fix share sheet

THISISDINOSAUR committed:
Fix root bookmark folder title

THISISDINOSAUR committed:
Fix section title showing on sub folders

THISISDINOSAUR committed:
Add edit text field clear icons

THISISDINOSAUR committed:
Rename linkImage on BookmarkCell

THISISDINOSAUR committed:
Update folder selected tick

THISISDINOSAUR committed:
Add contains methods

THISISDINOSAUR committed:
Fix saveBookmarkActivity

THISISDINOSAUR committed:
Clean up various todos

THISISDINOSAUR committed:
Fix home favorites view

THISISDINOSAUR committed:
Fix favorites overlay on ipad

THISISDINOSAUR committed:
Fix various actions on FavoritesHomeViewSectionRenderer and remove unused functions

THISISDINOSAUR committed:
Fix favorite editing from the home screen

THISISDINOSAUR committed:
Add favicons to edit bookmarks view

THISISDINOSAUR committed:
Fix ability to add a folder to itself (or a child of itself)

THISISDINOSAUR committed:
Change coredata contains methods to run on view context, since some
things that use them require a level of expediency

THISISDINOSAUR committed:
Fix browsing menu

THISISDINOSAUR committed:
Fix long press bookmark actions

THISISDINOSAUR committed:
Delete unused EditBookmarkAlert

THISISDINOSAUR committed:
Update project file

THISISDINOSAUR committed:
Update bookmark all tabs method

THISISDINOSAUR committed:
Fix auto selecting current folder when creating new folder

THISISDINOSAUR committed:
Add confirm alert when deleting a folder with contents

THISISDINOSAUR committed:
Set delete rule on folder children to cascade

THISISDINOSAUR committed:
Fix edit button not appearing after adding a folder to an empty subfolder

THISISDINOSAUR committed:
Change separator inset based on folder depth

THISISDINOSAUR committed:
Fix cell seperator colour

THISISDINOSAUR committed:
Set edit pages to start with title fields active

THISISDINOSAUR committed:
Fix tableview spacing on sub folders

THISISDINOSAUR committed:
Fix arranging folders on sub folder screens

THISISDINOSAUR committed:
Fix tab bar appearance issue

THISISDINOSAUR committed:
Change edit favorite/bookmark alert behaviour to allow going back to the
bookmarks page

THISISDINOSAUR committed:
Fix browsing menu remove favorite removing bookmark if url is both a favorite
and a bookmark

THISISDINOSAUR committed:
Stop empty folders from being added

THISISDINOSAUR committed:
Stop from adding bookmarks with empty titles or urls

THISISDINOSAUR committed:
Add save/dismiss on keyboard return to edit bookmark screen

THISISDINOSAUR committed:
Add save/dismiss on keyboard return to edit folder screen

THISISDINOSAUR committed:
Add url validation to edit bookmark/favorites screens

THISISDINOSAUR committed:
Readd display title to bookmarks

THISISDINOSAUR committed:
Fix pressing done on keyboard not animating navVC.pop

samsymons committed:
Utilize the RunLoop as a way to avoid database deadlocks.

samsymons committed:
Fix formatting in DateExtensionTests.

miasma13 committed:
Remove provisioning profile specifier from project as it is set by configuration file

THISISDINOSAUR committed:
Fix spacing of edit button and text clear buttons

THISISDINOSAUR committed:
Add separator between text fields on bookmark edit view

THISISDINOSAUR committed:
Fix number of files label disappearing for folders with long titles

THISISDINOSAUR committed:
Fix bookmark cell accessory styling

THISISDINOSAUR committed:
Add separator view to bookmark cells

THISISDINOSAUR committed:
Update cell spacing

THISISDINOSAUR committed:
Change bookmark cell spacing when editing

THISISDINOSAUR committed:
Add new empty state for sub folders

THISISDINOSAUR committed:
Merge branch 'develop' into feature/elle/bookmark-folders

THISISDINOSAUR committed:
Fix squiffy layout when swiping to delete

THISISDINOSAUR committed:
Fix existing issue with bookmarkviewcontroller editing when a cell is swiped

THISISDINOSAUR committed:
Fix toolbar item colours in dark mode

THISISDINOSAUR committed:
Fix bug where edit button appears when it shouldn't

THISISDINOSAUR committed:
Fix edit bookmark action sometimes editing favourite instead

THISISDINOSAUR committed:
Make bookmarks use its own persistent store

THISISDINOSAUR committed:
Update widgets to use core data store

THISISDINOSAUR committed:
Fix widget updating and favicon update on bookmark actions

THISISDINOSAUR committed:
Merge branch 'sam/core-data-deadlock-fixes' into feature/elle/bookmark-folders

THISISDINOSAUR committed:
Change to using singleton for BookmarksCoreDataStorage

THISISDINOSAUR committed:
Resolve various threading issues

THISISDINOSAUR committed:
Change browsing menu builder to be async, rather than abusing dispatch groups

THISISDINOSAUR committed:
Resolve potential threading issue with context creation

THISISDINOSAUR committed:
fix warning

THISISDINOSAUR committed:
Big coredata refactor to fix widgets and lay groundwork for migration
and generally increase thread safety

THISISDINOSAUR committed:
Add migration from user defaults

THISISDINOSAUR committed:
Add pixels

THISISDINOSAUR committed:
Resolve issue where widget would load bookmarks store multiple times

THISISDINOSAUR committed:
Make private contexts temporary

THISISDINOSAUR committed:
Add add folder option from edit bookmark screen and change how edit
screens are presented

THISISDINOSAUR committed:
Change done and edit button behaviours on main bookmark screen

THISISDINOSAUR committed:
Fix empty state subfolder cell background

THISISDINOSAUR committed:
Fix edit cell colours

THISISDINOSAUR committed:
Change create folder behaviour from edit bookmark view to select the newly created folder

THISISDINOSAUR committed:
Fix cell selection bug

THISISDINOSAUR committed:
Fix cell contents not being retained

THISISDINOSAUR committed:
Turn off autocorrect on url fields

THISISDINOSAUR committed:
Fix potential fatal error when saving, due to cell not existing

THISISDINOSAUR committed:
Fix intermittent crash on folder creation

THISISDINOSAUR committed:
Refactor bookmark item details datasources

THISISDINOSAUR committed:
Remove unused bookmark store

THISISDINOSAUR committed:
Revert "Remove unused bookmark store"

This reverts commit 1803b7443689e23499f0a9ae01b59418814048ca.

THISISDINOSAUR committed:
Move strings to UserText

THISISDINOSAUR committed:
Remove unused old data store from bookmarks manager

THISISDINOSAUR committed:
Remove unnecessary PresentableBookmarkItem

THISISDINOSAUR committed:
Refactor BookmarksDataSource

THISISDINOSAUR committed:
Inject bookmarks manager into data sources

THISISDINOSAUR committed:
Remove unused things from BookmarksSearch

THISISDINOSAUR committed:
Make injected bookmarkManagers private

THISISDINOSAUR committed:
Remove bookmark cell creation duplication

THISISDINOSAUR committed:
Inject bookmarks manager into favoritesHomeViewSectionRenderer

THISISDINOSAUR committed:
Remove unused todo

THISISDINOSAUR committed:
Inject bookmakrsmanager into SaveBookmarkActivity

THISISDINOSAUR committed:
Move strings to UserText

THISISDINOSAUR committed:
Give bookmark editing/saving/etc methods callbacks

THISISDINOSAUR committed:
Add errors to BookmarksCoreDataStorage

THISISDINOSAUR committed:
Remove commented out code

THISISDINOSAUR committed:
Update debug view controller with new bookmarks

THISISDINOSAUR committed:
Merge in develop

THISISDINOSAUR committed:
Remove AppURLs from bookmarkObject protocol

THISISDINOSAUR committed:
Minor clean up to migrate method

THISISDINOSAUR committed:
Fix core data error type

THISISDINOSAUR committed:
Fix accidental project file commits

THISISDINOSAUR committed:
Fix swift lint errors

THISISDINOSAUR committed:
Fix swift lint warnings

THISISDINOSAUR committed:
Revert accidental submodule change

THISISDINOSAUR committed:
Update bookmarkSearch tests

THISISDINOSAUR committed:
Remove unnecessary closures in BookmarksSearchTests

THISISDINOSAUR committed:
Merge branch 'develop' into feature/elle/bookmark-folders

# Conflicts:
#	Core/Pixel.swift

bwaresiak commented on DuckDuckGo/BookmarksSearch.swift:
```diff
@@ -19,34 +19,49 @@
+    }
+    
+    func bookmarksAndFavorites(completion: @escaping ([Bookmark]) -> Void) {
+        allBookmarksAndFavoritesShallow(completion: completion)
```
Why 'shallow'? Comment there suggests it is only used for deletion.

bwaresiak replied:
Also note how this method is called for every character typed - be it in url field or in bookmarks search - this can lead to performance issues.

THISISDINOSAUR replied:
Good point RE shallow, I can see how that's misleading. What do you think of replacing it with flat, since that's what I actually wanted to convey?
Re performance, my strategy was to try the simplest solution first and then test to see if the performance was okay, (which it seemed to be when I did my scale tests), to minimise the chance of inadvertently changing the search algorithm.
Do you have an opinion on what the best way to improving the performance might be? I can think of a few options, but I don't love any of them.

bwaresiak replied:
Yes, naming it Flat should bring clarity.

Idea that comes to my mind: we could create transactional-like object to manage the state of search:

1. User starts "operation" (url typing, search in bookmarks) that can lead to multiple searchers.
2. We spawn 'BookmarksCachingSearch' object.
3. That object retrieves current state and caches it - now we use search algo on these objects only.
4. When we complete "operation" (leave url/text field) we discard the `BookmarksCachingSearch` object.

Assumption is that state is not changed between the searches so we don't need to re-fetch data. Object lifetime is tied to the cache so it is easy to use.

bwaresiak commented on Core/BookmarksCoreDataStorage.swift:
```diff
@@ -0,0 +1,722 @@
+        
+        func updateParentAndSave(parent: BookmarkFolderManagedObject) {
+            item.parent = parent
+            parent.addToChildren(item)
```
This is not needed (and I've seen this pattern in other places) - since relationships are bi-directional Core Data automatically manages graph consistency, so setting relation from one side is enough ( `item.parent = parent` )

THISISDINOSAUR replied:
I didn't think it was but was seeing some issue a long time ago related to these saves. I think that was an unrelated problem though that I resolved so I'll change these

bwaresiak commented on Core/BookmarksCoreDataStorage.swift:
```diff
@@ -0,0 +1,722 @@
+        let privateContext = getTemporaryPrivateContext()
+        privateContext.perform {
+
+            let mo = privateContext.object(with: folderID)
```
We should not be using `object(with:)` but rather `existingObject(with:)` as the former one always returns object - even if it has ben removed from the store, which can result in the crash if we reach try to unfault such object.

bwaresiak commented on DuckDuckGo/BookmarksSearch.swift:
```diff
@@ -92,21 +107,29 @@ class BookmarksSearch {
-        if sortByRelevance {
-            finalResult = finalResult.sorted { $0.score > $1.score }
+        bookmarksStore.bookmarksAndFavorites { bookmarks in
+            let results: [ScoredBookmark] = bookmarks.map {
```
Hm, it feels a bit awkward that search ignores folders.

THISISDINOSAUR replied:
Yeah, we discussed it and decided to exclude them, which is what Android did.

bwaresiak replied:
I understand, thanks for confirmation!

bwaresiak commented on Core/BookmarksCoreDataStorage.swift:
```diff
@@ -0,0 +1,722 @@
+        viewContext.perform {
+    
+            let fetchRequest = NSFetchRequest<BookmarkManagedObject>(entityName: Constants.bookmarkClassName)
+            fetchRequest.predicate = NSPredicate(format: "url == %@ AND isFavorite == false", url as CVarArg)
```
To make this less error-prone and avoid writing property names by hand this could be written like:

`fetchRequest.predicate = NSPredicate(format: "%K == %@ AND %K == false", #keyPath(BookmarkManagedObject.url), url as NSURL, #keyPath(BookmarkManagedObject.isFavorite))`

You also can cast to actual Foundation object instead of CVarArg.

bwaresiak commented on Core/BookmarksCoreDataStorage.swift:
```diff
@@ -0,0 +1,722 @@
+                completion(false)
+                return
+            }
+            completion(results.count > 0)
```
If you are only interested in the count, then call `viewContext.count(for: fetchRequest)` instead of `viewContext.fetch(fetchRequest)`. Moreover, if it's only about existence (count > 0) then you can add limit to the fetch request: `request.fetchLimit = 1`

bwaresiak commented on DuckDuckGo/TabSwitcherViewController.swift:
```diff
@@ -224,19 +224,25 @@ class TabSwitcherViewController: UIViewController {
             }
         }
         
+        group.wait()
```
This will cause deadlock as bookmarksManager is tied to main thread and this is also invoked on main thread -  completion block of `bookmarksManager.contains` will be always waiting for execution on the thread that is currently blocked.

These kind of deadlocks (hard to foresee in a perfectly reasonably written code)  are the reason why it is advocated to return completion blocks on private queues unless there is a good reason not to, or that class is directly tied to main thread and it is made obvious to the developer (e.g. UIKit classes, animations, possibly ViewModels).

Ideally Bookmarks Manager should be thread-agnostic - but I understand how dispatching all completion blocks back to main thread could be a pain. What do you think?

THISISDINOSAUR replied:
I'm really not sure on this one, in other cases I just ended up converting previously synchronous functions to async (e.g. building the browsing menu). This one will be a bit more of a PITA to do I think, but I don't have a better idea, do you?

bwaresiak replied:
I think that answer to https://github.com/duckduckgo/iOS/pull/1014#discussion_r770510085 may be helpful - if we were to go with the "completion on private queue only" then this issue should disappear, but you'd need to add `DispatchQueue.main` calls in other places. 

Alternatively, we could go with runloop juggling (ugh) or change this method to also have completion and invoke it on background queue.

THISISDINOSAUR replied:
So you mean keep the dispatch group, but ensure all the bookmakarks manager calls complete on a private queue?

bwaresiak replied:
Yes, this way there is no way for it to deadlock in such manner.

THISISDINOSAUR replied:
I've implemented and pushed a recursive function for this, but it means the tabs aren't necessarily saved in order. LMK if you can think of an easy way to resolve this

THISISDINOSAUR replied:
Okay, ignore me, It was a very simple fix, it's late here...

bwaresiak commented on Core/BookmarksCoreDataStorage.swift:
```diff
@@ -0,0 +1,722 @@
+    public func update(folderID: NSManagedObjectID,
+                       newTitle: String,
+                       newParentID: NSManagedObjectID,
+                       completion: BookmarkItemUpdatedCompletion? = nil) {
```
One nitpick about completion handlers in this class: sometimes these will be invoked on main thread, sometimes on private queue of a private context - this isn't really consistent and may cause issues, as we cannot make any assumptions about the API.

THISISDINOSAUR replied:
What would your preference be given objects being somewhat tied to threads?

bwaresiak replied:
I don't think any objects that are retrieved from this API are tied to any particular thread - or am missing something?

THISISDINOSAUR replied:
I mean with e.g. `public func bookmark(forURL url: URL, completion: @escaping (Bookmark?) -> Void)`, it's completion will be invoked on the main thread, which sort of makes sense to me since the object is from the view context. Doesn't the fact that it's from the view context (and e.g. if I tried to check the parent or child and it's faulted on a background thread it could crash) mean we have to make assumptions about the API?

bwaresiak replied:
Sorry, I got totally confused looking at the API. My mind got the impression that `Bookmark` is not a managed object - and thus it is not tied to any thread.

Now I have few trains of thought:
1. We could make API more explicit about what is returned: seeing `BookmarkManagedObject` in the completion would make people think about thread affiliation of particular instances. It would also disclose the information about inner implementation, which in this case is not bad as API has limitation around threading related to Core Data. I know it's not ideal as `Bookmark` abstraction is elegant, but at the same time it does not warn the user of this API about its limitations (I've assumed it's a thread-safe struct 🤦‍♂️and was ready to just pass it between threads... )
2. For methods that return object IDs in completion, I'd consider if we want to make it also return on the Main Thread, to keep API consistent. That would tie the whole object to Main Thread, but not in the performance-impairing way.
3. If users of this API understand relation to Main thread (based on above two points), then it is less likely any locking they implement won't take that into account - DispatchGroup from the above would have to be transformed to be async call based on that.

It would make this object fall in the "there is a good reason not use the private queue for completions" bucket. Does that makes sense to you?

THISISDINOSAUR replied:
Yeah, that does make sense. Agreed on 1, I'd rather we had the abstraction, but, well, we just don't as long as we're using core data properly. At least at the UI layer etc we can still have it be more generic 
Do you think I should change BookmarksManager to return a `BookmarkManagedObject` instead too? 
2 also makes sense, but it still wouldn't feel super obvious from the outside to me. How would you feel about something in the completion type alias name to indicate the thread?

bwaresiak replied:
Yes to all of your points, we should make it as clear as possible. :)

bwaresiak commented on Core/BookmarksCoreDataStorage.swift:
```diff
@@ -0,0 +1,722 @@
+        
+        if sender.name == Constants.privateContextName {
+            viewContext.perform { [weak self] in
+                self?.viewContext.mergeChanges(fromContextDidSave: notification)
```
Very good. 😄

bwaresiak reviewed (changes requested):
Great job @THISISDINOSAUR it really works well. I've dropped some comment in the code, will be giving it another pass Tomorrow. 😄

THISISDINOSAUR committed:
Rename shallow to flat for clarity

THISISDINOSAUR committed:
Remove unnecessary add to/remove from children calls

THISISDINOSAUR committed:
Change calls to object(with:) to existingObject(with:)

THISISDINOSAUR committed:
Change predicates to use keypaths for params

THISISDINOSAUR committed:
Change contains function to use count instead of fetch

THISISDINOSAUR committed:
Make it explicit that core data/bookmarkManager functions return managed objects

THISISDINOSAUR committed:
Make completions that are called with a NSManagedObjectID complete on main thread

THISISDINOSAUR committed:
Rename completions to make it clear which thread they return on

THISISDINOSAUR committed:
Introduce typealias for contains completion, to clarify which thread it is called on

THISISDINOSAUR committed:
Fix lint warnings

THISISDINOSAUR committed:
Make bookmark all tabs async

THISISDINOSAUR committed:
Add missing

THISISDINOSAUR committed:
Fix bookmarking all tabs not being in the right order

THISISDINOSAUR committed:
Make bookmark search caching

THISISDINOSAUR committed:
Change BookmarksViewController search caching object to be initialised when
text field is active, not when view controller is loaded, to avoid sometimes
loading unnecessarily

THISISDINOSAUR committed:
Update BookmarksSearch tests

THISISDINOSAUR committed:
Fix lint error

THISISDINOSAUR committed:
Add translations

THISISDINOSAUR committed:
Revert "Add translations"

This reverts commit 42800d80651f64cb92ed68dc8c581ad84391f91e.

THISISDINOSAUR committed:
Add translations

bwaresiak commented on DuckDuckGo/BookmarksCachingSearch.swift:
```diff
@@ -92,21 +124,28 @@ class BookmarksSearch {
+    func search(query: String, sortByRelevance: Bool = true, completion: @escaping ([Bookmark]) -> Void) {
         guard hasData else {
-            return []
+            completion([])
```
I'm curious if this will happen - debug pixel maybe?

THISISDINOSAUR replied:
It should just happen if the user hasn't added anything, it's not checking the cache is populated or anything like that

THISISDINOSAUR replied:
Although there'd be a decent argument for getting rid of this check as unnecessary, but I'm somewhat indifferent about it

THISISDINOSAUR committed:
Fix share sheet options

THISISDINOSAUR committed:
Add explanation comment

bwaresiak commented on DuckDuckGo/BookmarksCachingSearch.swift:
```diff
@@ -18,35 +18,67 @@
-    init(bookmarksStore: BookmarkStore = BookmarkUserDefaults()) {
+    init(bookmarksStore: BookmarksSearchStore = BookmarksCoreDataStorage.shared) {
         self.bookmarksStore = bookmarksStore
+        loadCache()
```
When I open Bookmarks and search for something, cache is not reset when leaving the text field - thus searching after deleting bookmark still returns it.

bwaresiak commented on DuckDuckGo/AddOrEditBookmarkViewController.swift:
```diff
@@ -0,0 +1,158 @@
+import Core
+import CoreData
+
+class AddOrEditBookmarkViewController: UIViewController {
```
When Editing bookmarks I get some layout constraints warnings (running on iPhone SE 1st gen).

bwaresiak commented on Core/BookmarksCoreDataStorage.swift:
```diff
@@ -0,0 +1,762 @@
+public typealias BookmarkItemDeletedBackgroundThreadCompletion = ((Bool, BookmarksCoreDataStorageError?) -> Void)
+public typealias BookmarkItemUpdatedBackgroundThreadCompletion = ((Bool, BookmarksCoreDataStorageError?) -> Void)
+public typealias BookmarkItemIndexUpdatedBackgroundThreadCompletion = ((Bool, BookmarksCoreDataStorageError?) -> Void)
+public typealias BookmarkConvertedBackgroundThreadCompletion = ((Bool, BookmarksCoreDataStorageError?) -> Void)
```
This looks pretty clear :)

bwaresiak commented on Core/BookmarksCoreDataStorage.swift:
```diff
@@ -0,0 +1,762 @@
+    public func allBookmarksAndFavoritesFlat(completion: @escaping ([BookmarkManagedObject]) -> Void) {
+        viewContext.perform { [weak self] in
+        
+            let fetchRequest: NSFetchRequest<BookmarkManagedObject> = BookmarkManagedObject.fetchRequest()
```
I'd consider adding parameter to configure whether we want to obtain bookmarks objects that are already prefetched (non-faults) to improve performance of search.

THISISDINOSAUR replied:
Sorry can you clarify what you mean?

bwaresiak replied:
By default `fetchRequest` does not populate content of objects from the DB - so when you iterate over the result set (e.g. in search logic) it causes N (where N is number of objects) DB queries - one for each object. Instead, you can configure [`returnsObjectsAsFaults`](https://developer.apple.com/documentation/coredata/nsfetchrequest/1506756-returnsobjectsasfaults) to make things more efficient. It would be good to have that as an option here.

THISISDINOSAUR replied:
Should it be an option? Sounds like I should just disable faulting for this method tbh

bwaresiak replied:
I'd say yes - to not bind this API with certain usage. But I don't have strong opinion, you could just set that property to false here to improve performance of current use cases.

bwaresiak commented on Core/BookmarksCoreDataStorage.swift:
```diff
@@ -0,0 +1,762 @@
+        guard let folder = cachedReadOnlyTopLevelBookmarksFolder else {
+            return []
+        }
+        return folder.children?.array as? [BookmarkItemManagedObject] ?? []
```
Not for now, but generally: above is a strong reference (and it has to be to serve its purpose) but that may lead to Core Data keeping objects in memory - need to remember that if this turns out to be a problem someday.

bwaresiak commented on DuckDuckGo/BookmarkSectionDataSources.swift:
```diff
@@ -0,0 +1,251 @@
+            if count == 1 {
+                messageString = UserText.deleteBookmarkFolderAlertMessageSingular
+            } else {
+                messageString = UserText.deleteBookmarkFolderAlertMessagePlural
```
Shouldn't this use strings dict to properly translate plural form? See `TabSwitcherViewController:refreshTitle()` for reference.

Also, if I have:
Bookmarks -> Folder A (one subfolder) -> Folder B (10 subfolders) message for removal of 'Folder A' will only warn me about removing one item, not a bunch of these from a subtree.

bwaresiak commented on DuckDuckGo/BookmarksViewController.swift:
```diff
@@ -21,26 +21,29 @@ import UIKit
 class BookmarksViewController: UITableViewController {
 
+    @IBOutlet weak var addFolderButton: UIBarButtonItem!
     @IBOutlet weak var editButton: UIBarButtonItem!
```
We probably should disable Add Folder & Edit buttons when in search mode (keyboard can be invisible when using external one).

bwaresiak reviewed (commented):
Great job, I've left few comments inline but overall it looks really solid and work really well. I went through the testing scenarios (including migration) and nothing looks broken, well done. :)

Feel free to merge when you're done.

THISISDINOSAUR committed:
Fix search cache not being refreshed if edits have occurred

THISISDINOSAUR committed:
Hide add folder and edit buttons when searching

THISISDINOSAUR committed:
Fix lint error

THISISDINOSAUR committed:
Fix folder delete dialog not taking into account sub folder children

THISISDINOSAUR committed:
Fix Failed to set user defined inspected property on BookmarksViewController error

THISISDINOSAUR committed:
Fix Failed to set user defined inspected property on edit folder view

THISISDINOSAUR committed:
Resolve autolayout errors

THISISDINOSAUR committed:
Fix IB object having random class

THISISDINOSAUR committed:
Resolve autolayout constraint errors

THISISDINOSAUR committed:
Disable faulting when fetching all bookmarks flat

THISISDINOSAUR committed:
Fix lint error
